crates/ika-core/src/dwallet_mpc/mpc_session.rs

```diff
@@ -754,43 +795,85 @@ impl DWalletMPCSession {
-                        )
-                        .is_ok()
-                        && self.received_more_messages_since_last_advance)
-                {
-                    ReadyToAdvanceCheckResult {
+                if self.current_round == 1 {
+                    Ok(ReadyToAdvanceCheckResult {
                         is_ready: true,
                         malicious_parties: vec![],
+                    })
+                } else if (self
+                    .weighted_threshold_access_structure
+                    .is_authorized_subset(
+                        &self
+                            .serialized_full_messages
+                            // Check if we have the threshold of messages for the previous round
+                            // to advance to the next round.
+                            .get(&(self.current_round - 1))
+                            .unwrap_or(&HashMap::new())
+                            .keys()
+                            .cloned()
+                            .collect::<HashSet<PartyID>>(),
+                    )
+                    .is_ok()
+                    && self.received_more_messages_since_last_advance
+                    && self.agreed_mpc_protocol.is_some())
+                {
+                    let epoch_store = self.epoch_store()?;
+                    let protocol_consensus_rounds_delay = epoch_store
+                        .protocol_config()
+                        .consensus_rounds_delay_per_mpc_protocol
+                        .get(&self.agreed_mpc_protocol.clone().unwrap())
+                        .unwrap_or(&HashMap::new())
+                        .get(&(self.current_round as u64))
+                        .unwrap_or(&0u64)
+                        .clone();
+                    if self.consensus_rounds_since_quorum_reached
+                        >= protocol_consensus_rounds_delay as usize
+                    {
+                        info!(
+                            ?protocol_consensus_rounds_delay,
+                            ?self.consensus_rounds_since_quorum_reached,
+                            ?self.current_round,
+                            ?self.agreed_mpc_protocol,
+                            ?self.session_id,
+                            "Quorum reached for MPC session and delay passed, advancing to next round",
+                        );
+                        self.consensus_rounds_since_quorum_reached = 0;
+                        Ok(ReadyToAdvanceCheckResult {
+                            is_ready: true,
+                            malicious_parties: vec![],
+                        })
+                    } else {
+                        info!(
+                            ?protocol_consensus_rounds_delay,
+                            ?self.consensus_rounds_since_quorum_reached,
+                            ?self.current_round,
+                            ?self.agreed_mpc_protocol,
```

Comment:
please log in both cases the number of messages for this round we have right now

Reply:
also in store message